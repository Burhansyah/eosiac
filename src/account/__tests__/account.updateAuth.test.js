const cloneDeep = require(`lodash/cloneDeep`)
const Account = require(`../index`)
const setup = require(`./setup`)

const {env, accountConfig, existingAccount} = setup

const existingActivePermissions = {
    threshold: 2,
    permissions: [
        `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
        `EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32`,
        `dapptoken@eosio.code 1`,
        `wait@600 1`,
    ],
}

describe(`account.updateAuth`, () => {
    it(`throws when account does not exist yet`, async () => {
        expect.assertions(1)
        const a = new Account(`test`, accountConfig)
        a.currentState = {}
        expect(a.updateAuth({env})).rejects.toThrow(/does not exist/i)
    })

    it(`returns no actions to run when used on same unordered permissions`, async () => {
        expect.assertions(1)
        const $accountConfig = cloneDeep(accountConfig)
        $accountConfig.auth.active.permissions = [
            `EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32`,
            `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
            `dapptoken@eosio.code 1`,
            `wait@600 1`,
        ]
        const a = new Account(`test`, $accountConfig)
        a.currentState = cloneDeep(existingAccount)
        const actions = await a.updateAuth({env})
        expect(actions).toBeFalsy()
    })

    it(`returns updateauth action when used with new permission`, async () => {
        expect.assertions(1)
        const $accountConfig = cloneDeep(accountConfig)
        $accountConfig.auth = {
            owner: {
                permissions: [
                    `EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi`,
                    `EOS7uSfs552n9hN1U2vidQybeP6dqx2V4qLEB5BdtEGPM3kyhGDnd`,
                ],
            },
            active: {
                threshold: 3,
                permissions: [
                    `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
                    `EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32`,
                    `dapptoken@eosio.code 1`,
                    `wait@600 1`,
                ],
            },
            newnewnew: {
                threshold: 2,
                permissions: [`test@active`, `wait@1200 1`],
            },
        }
        const a = new Account(`test`, $accountConfig)
        a.currentState = cloneDeep(existingAccount)
        const actions = await a.updateAuth({env})
        expect(actions).toMatchSnapshot()
    })

    it(`returns updateauth and linkauth action when used with new permission`, async () => {
        expect.assertions(1)
        const $accountConfig = cloneDeep(accountConfig)
        $accountConfig.auth = Object.assign($accountConfig.auth, {
            xtransfer: {
                parent: `active`,
                permissions: [`test@active`],
                links: [`eosio.token@transfer`],
            },
            active: existingActivePermissions,
        })
        const a = new Account(`test`, $accountConfig)
        a.currentState = cloneDeep(existingAccount)
        const actions = await a.updateAuth({env})
        expect(actions).toMatchSnapshot()
    })

    it(`skips linkauth action when already linked`, async () => {
        expect.assertions(1)
        const $accountConfig = cloneDeep(accountConfig)
        $accountConfig.auth = Object.assign($accountConfig.auth, {
            xtransfer: {
                parent: `active`,
                permissions: [`test@active`],
                links: [`eosio.token@transfer`],
            },
            active: existingActivePermissions,
        })
        const a = new Account(`test`, $accountConfig)
        a.currentState = cloneDeep(existingAccount)
        a.currentState.linkedPermissions = [
            {contract: `eosio.token`, action: `transfer`, permission_name: `xtransfer`},
        ]
        const actions = await a.updateAuth({env})
        expect(actions).toMatchSnapshot()
    })

    it(`returns unlinkauth action when linked but should not be anymore`, async () => {
        expect.assertions(1)
        const $accountConfig = cloneDeep(accountConfig)
        $accountConfig.auth = Object.assign($accountConfig.auth, {
            xtransfer: {
                parent: `active`,
                permissions: [`test@active`],
            },
            active: existingActivePermissions,
        })
        const a = new Account(`test`, $accountConfig)
        a.currentState = cloneDeep(existingAccount)
        a.currentState.linkedPermissions = [
            {contract: `eosio.token`, action: `transfer`, permission_name: `xtransfer`},
        ]
        const actions = await a.updateAuth({env})
        expect(actions).toMatchSnapshot()
    })
})
