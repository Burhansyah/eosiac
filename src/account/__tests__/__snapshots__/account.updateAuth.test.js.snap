// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account.updateAuth returns updateauth action when used with new permission 1`] = `
Array [
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "owner",
      },
    ],
    "data": Object {
      "account": "test",
      "auth": Object {
        "accounts": Array [],
        "keys": Array [
          Object {
            "key": "EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi",
            "weight": 1,
          },
          Object {
            "key": "EOS7uSfs552n9hN1U2vidQybeP6dqx2V4qLEB5BdtEGPM3kyhGDnd",
            "weight": 1,
          },
        ],
        "parent": "",
        "threshold": 1,
        "waits": Array [],
      },
      "parent": "",
      "permission": "owner",
    },
    "name": "updateauth",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "owner",
      },
    ],
    "data": Object {
      "account": "test",
      "auth": Object {
        "accounts": Array [
          Object {
            "permission": Object {
              "actor": "dapptoken",
              "permission": "eosio.code",
            },
            "weight": 1,
          },
        ],
        "keys": Array [
          Object {
            "key": "EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x",
            "weight": 1,
          },
          Object {
            "key": "EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32",
            "weight": 1,
          },
        ],
        "parent": "owner",
        "threshold": 3,
        "waits": Array [
          Object {
            "wait_sec": 600,
            "weight": 1,
          },
        ],
      },
      "parent": "owner",
      "permission": "active",
    },
    "name": "updateauth",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "owner",
      },
    ],
    "data": Object {
      "account": "test",
      "auth": Object {
        "accounts": Array [
          Object {
            "permission": Object {
              "actor": "test",
              "permission": "active",
            },
            "weight": 1,
          },
        ],
        "keys": Array [],
        "parent": "owner",
        "threshold": 2,
        "waits": Array [
          Object {
            "wait_sec": 1200,
            "weight": 1,
          },
        ],
      },
      "parent": "owner",
      "permission": "newnewnew",
    },
    "name": "updateauth",
  },
]
`;
